eJWST prelaunch

jespinosaar opened the pull request:
Dear Astroquery team:

As James Webb Space Telescope (JWST) will be launched on November, we have developed a module to access future ESA JWST Archive using astroquery. I am creating this draft pull request in advance to prepare our module to be integrated in astroquery and inform you about the necessity of merging this pull request at an agreed date with you (before November, I will confirm you a specific date in a few weeks) and, if possible, agree in releasing a new version of astroquery.

This is still ongoing work, no major changes are expected, but you can have a look at this PR if you want and comment anything you think is useful.

Thanks in advance!

Kind regards,

Javier Espinosa

pep8speaks commented:
Hello @jespinosaar! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

























* In the file [`astroquery/jwst/core.py`](https://github.com/astropy/astroquery/blob/b7a1efac761e0a3f450889e315476b9fc453b34e/astroquery/jwst/core.py):

> [Line 80:83](https://github.com/astropy/astroquery/blob/b7a1efac761e0a3f450889e315476b9fc453b34e/astroquery/jwst/core.py#L80): [E703](https://duckduckgo.com/?q=pep8%20E703) statement ends with a semicolon
> [Line 1573:9](https://github.com/astropy/astroquery/blob/b7a1efac761e0a3f450889e315476b9fc453b34e/astroquery/jwst/core.py#L1573): [E722](https://duckduckgo.com/?q=pep8%20E722) do not use bare 'except'

* In the file [`astroquery/jwst/data_access.py`](https://github.com/astropy/astroquery/blob/b7a1efac761e0a3f450889e315476b9fc453b34e/astroquery/jwst/data_access.py):

> [Line 24:1](https://github.com/astropy/astroquery/blob/b7a1efac761e0a3f450889e315476b9fc453b34e/astroquery/jwst/data_access.py#L24): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (3)

keflavich commented:
Thanks for the heads up.  I certainly acknowledge the importance of a timely deployment for the JWST archive.  Though I suspect we'll have a little while post-launch before there are any data to serve, no?

Is there any redundant code here with the other TAPPlus in astroquery?  Can that be reused instead of copied?

jespinosaar commented:
Hi @keflavich, thanks for your comments. Yes, we will have to wait some time before we have data to serve and, for this reason, we would like to have this PR merged before, to be ready when this happens.

I think I can refactor code that was copied from TAPPlus login window, as we needed a new one with an additional field. I will revise it together with new classes. Thanks!

eerovaher commented:
I don't know about the broken tests, but I did notice that this pull request would introduce quite a lot of Python2-related code to `astroquery`, which works against #1870. For example, `six` has been removed from `astroquery` and is no longer a direct requirement, but this pull request seems to rely on `six` quite a lot in `astroquery/esa/jwst/token.py`.

jespinosaar commented:
> I don't know about the broken tests, but I did notice that this pull request would introduce quite a lot of Python2-related code to `astroquery`, which works against #1870. For example, `six` has been removed from `astroquery` and is no longer a direct requirement, but this pull request seems to rely on `six` quite a lot in `astroquery/esa/jwst/token.py`.

astroquery/esa/jwst/token.py file has been removed, as, in the end, it was not necessary anymore. Which other classes do you think we should modify to remove Python2-related code?

Anyway, we still have the issue with SIMBAD...

bsipocz commented:
> I have a question, JWST launch will be in December, but we would like to have this code merged at the end of November (no data will be accessible, of course). Would it be also possible to have an official astroquery release with this new module?

I'm planning to do a tagged release at the end of October to address the JPL API changes. But also happy to tag one once this is in, before the launch. 
(We will also have dev version release on pypi sooner (e.g. can have that with the merge commit), but I agree that this module will be a significant addition that deserves a tagged release that ends up on conda-forge, etc, too.)

bsipocz commented:
Re unrelated CI failures: I can't recall seeing this one, but a rebase should certainly resolve the issue.

jespinosaar commented:
Hi @bsipocz ,

I have updated the code according to your comments. I have also simplified the code as I told you, removing the async methods and including the tag "async_job" to execute a sync or async method, I hope it is ok for you. I have finally rebased the code, and it seems that the issue with oldest dependencies is still there. Could you please have a look at it? Thanks in advance!

jespinosaar commented:
It seems I have also a problem in the documentation...

bsipocz commented:
> It seems I have also a problem in the documentation...

Yes, we're experiencing some issues with upstream astropy. I'm tending those, cut 0.4.4 and then come back to this (and will do 0.4.5 as we agreed above)

jespinosaar commented:
> > It seems I have also a problem in the documentation...
> 
> Yes, we're experiencing some issues with upstream astropy. I'm tending those, cut 0.4.4 and then come back to this (and will do 0.4.5 as we agreed above)

Thanks @bsipocz !
Once everything is working nominally, when all the checks are passed, please do not forget we have a dummy TAP url that needs to be modified to the correct one before merging the code.

jespinosaar commented:
Hi @bsipocz , any update in the issues? thanks in advance!

jespinosaar commented:
Hi @bsipocz, @keflavich,  first of all, congratulations on version 0.4.4! It seems it has been a headache...

On the other hand, I have rebased again the code, but the issues are still there... Sorry for my insistence, but we would like to have this module merged by the end of the month. Could you please review it? Thanks a lot!

bsipocz commented on astroquery/esa/jwst/core.py:
```diff
@@ -0,0 +1,1231 @@
+    CAL_LEVELS = ['ALL', 1, 2, 3, -1]
+    REQUESTED_OBSERVATION_ID = "Missing required argument: 'observation_id'"
+
+    def __init__(self, tap_plus_handler=None, data_handler=None):
```
Maybe make `tap_plus_handler` and `data_handler` kwarg only, too?

jespinosaar replied:
Done

bsipocz commented on astroquery/esa/jwst/core.py:
```diff
@@ -0,0 +1,1231 @@
+        if data_handler is None:
+            self.__jwstdata = JwstDataHandler(
+                base_url=conf.JWST_DATA_SERVER)
+            # self.__jwstdata = self.__jwsttap;
```
is this commented out temporarily?

jespinosaar replied:
No, now it is removed

bsipocz commented on astroquery/esa/jwst/tests/test_jwsttap.py:
```diff
@@ -0,0 +1,1088 @@
+def plane_id_request(request):
+    try:
+        mp = request.getfixturevalue("monkeypatch")
+    except AttributeError:  # pytest < 3
```
we are not supporting this old pytests, so you can drop the try/except, here and below, too

jespinosaar replied:
Ok, done

bsipocz commented on astroquery/esa/jwst/tests/test_jwsttap.py:
```diff
@@ -0,0 +1,1088 @@
+def get_product_request(request):
+    try:
+        mp = request.getfixturevalue("monkeypatch")
+    except AttributeError:  # pytest < 3
```
we are not supporting this old pytests, so you can drop the try/except

jespinosaar replied:
Ok, done

bsipocz commented on astroquery/esa/jwst/tests/test_jwsttap.py:
```diff
@@ -0,0 +1,1088 @@
+    return mp
+
+
+planeids = "('00000000-0000-0000-879d-ae91fa2f43e2', "\
```
No need to reformat in general, but an a reminder: line lenghts up to 120 are allowed, no need to do line breaks in places where it doesn't necessarily help readibility.

jespinosaar replied:
I have finally reformatted to improve readibility.

bsipocz commented on astroquery/esa/jwst/tests/test_jwsttap.py:
```diff
@@ -0,0 +1,1088 @@
+                      frame='icrs')
+        with pytest.raises(ValueError) as err:
+            tap.query_region(sc)
+            print(err.value.args[0])
```
This print won't return anything even if the assert fails below, so can be cleaned up

jespinosaar replied:
Done

bsipocz commented on astroquery/esa/jwst/tests/test_jwsttap.py:
```diff
@@ -0,0 +1,1088 @@
+        # Test observation_id argument
+        with pytest.raises(ValueError) as err:
+            tap.query_region(sc, width=width, height=height, observation_id=1)
+        assert "observation_id must be string" in err.value.args[0]
```
would you test it with a valid string input, too?

jespinosaar replied:
Tested with valid inputs after all assertions

bsipocz commented on astroquery/esa/jwst/tests/test_jwsttap.py:
```diff
@@ -0,0 +1,1088 @@
+        assert "prod_type must be string" in err.value.args[0]
+
+        with pytest.raises(ValueError) as err:
+            tap.query_region(sc, width=width, height=height, prod_type='a')
```
please have a test for the valid input, too

jespinosaar replied:
Tested with valid inputs after all assertions

bsipocz commented on astroquery/esa/jwst/tests/test_jwsttap.py:
```diff
@@ -0,0 +1,1088 @@
+
+        with pytest.raises(ValueError) as err:
+            tap.query_region(sc, width=width, height=height,
+                             instrument_name='a')
```
again, please have a test with a valid input, too. And I might have missed a few kwargs where you test for invalid inputs only, it would be great to have a test for the valid ones, too. It's OK, to leave it for the remote tests though rather than mock it.

jespinosaar replied:
Tested with valid inputs after all assertions

bsipocz commented on astroquery/esa/jwst/tests/test_jwsttap.py:
```diff
@@ -0,0 +1,1088 @@
+        assert "proposal_id must be string" in err.value.args[0]
+
+        table = tap.query_region(sc, width=width, height=height)
+        assert len(table) == 3, \
+            "Wrong job results (num rows). Expected: %d, found %d" % \
+            (3, len(table))
```
this is a typical case where I feel a single line would be more readable, all would fit into <100 chars. (f-stings can also be used)

eerovaher replied:
Suppose we have a small test module that looks like this:
```python
from astropy.table.table_helpers import simple_table

def test():
    table = simple_table(2)
    assert len(table) == 3
```
Running it produces the following:
```
=================================== FAILURES ===================================
_____________________________________ test _____________________________________

    def test():
        table = simple_table(2)
>       assert len(table) == 3
E       assert 2 == 3
E        +  where 2 = len(<Table length=2>\n  a      b     c  \nint64 float64 str1\n----- ------- ----\n    1     1.0    c\n    2     2.0    d)

set4/test_solutions.py:5: AssertionError
=========================== short test summary info ============================
FAILED set4/test_solutions.py::test - assert 2 == 3
============================== 1 failed in 0.10s ==============================
```
With the custom assertion error message the output is:
```
=================================== FAILURES ===================================
_____________________________________ test _____________________________________

    def test():
        table = simple_table(2)
>       assert len(table) == 3, \
                "Wrong job results (num rows). Expected: %d, found %d" % \
                (3, len(table))
E       AssertionError: Wrong job results (num rows). Expected: 3, found 2
E       assert 2 == 3
E        +  where 2 = len(<Table length=2>\n  a      b     c  \nint64 float64 str1\n----- ------- ----\n    1     1.0    c\n    2     2.0    d)

set4/test_solutions.py:5: AssertionError
=========================== short test summary info ============================
FAILED set4/test_solutions.py::test - AssertionError: Wrong job results (num ...
============================== 1 failed in 0.11s ===============================
```
I don't see much difference between the outputs. Wouldn't it be simpler not to write the custom assertion error message or is there some benefit I'm not noticing?

bsipocz replied:
I myself don't write custom assert messages, and the generic part of astroquery or other libraries I'm more involved in don't have them. But I fully accept and support if a maintainer group of a module find those useful. So here I would totally leave it to @jespinosaar and others at ESA to have their preference. My comment was really just an example that the we don't need to stick with line lengths of 80 especially when the code could be made more readable and compact using a little longer lines, yet staying within 100-120.

jespinosaar replied:
Text has been reformatted

bsipocz commented on docs/esa/jwst.rst:
```diff
@@ -0,0 +1,751 @@
+ESA, and the Canadian Space Agency (CSA). Although radically different in
+design, and emphasizing the infrared part of the electromagnetic spectrum,
+JWST is widely seen as the successor to the Hubble Space Telescope (HST).
+The JWST observatory will consist of a deployable 6.6 meter passively cooled
```
maybe rephrase for the present tense from the future, after all this module will only be used when JWST is already operational

jespinosaar replied:
I have rephrased it, thanks for looking into this!

bsipocz reviewed (approved):
Some minor comments, none of which is a blocker for merge.

The main remaining convo from the previous review was around some code simplification where you raised the idea of setting async/sync with a kwarg pointing out that the decorator approach would not work as is with TapPlus. That's absolutely right, I'm sorry to missed that during review. 
And I don't think the kwarg approach is an API choice we want to make as the rest of the package is not using it. Hindsight we could just have done that rather than the duplicated methods, but I don't think it's realistic to change it now.


So, overall: let's merge this. All CI failures seems unrelated, let me try to clean those before hitting the button. And I believe @jespinosa, you may also need to change the URLs before merge? Or is that after, just right before launch time? Let me know.

jespinosaar commented:
Thanks a lot for your time, @bsipocz. Although they are not blockers, I will implement your latest comments. Let's see if we can have all the issues solved and ready to merge soon.

On the other hand, yes, we need to modify the URLs to use our operational TAP server before merging this code. There have been some delays in the release dates, as the DB clean up will finally happen during the second week of December and we need to put in production our TAP server with the cleaned DB, so we now expect to have this merged by mid-December, I will tell you a more specific date ASAP.

(P.S. even JWST launch has been delayed to no earlier than December 22nd).

bsipocz commented:
@jespinosaar - I have a slight preference to merge it a few (1-2) weeks before the tentative release date, even if it's non functional yet due to upstream server limitations. Just to make sure merging it won't be blocked by some unrelated issues, etc.

jespinosaar commented:
> @jespinosaar - I have a slight preference to merge it a few (1-2) weeks before the tentative release date, even if it's non functional yet due to upstream server limitations. Just to make sure merging it won't be blocked by some unrelated issues, etc.

Hi @bsipocz, @eerovaher,  your approach is understandable, we have been discussing it and yes, what we can do is merging this pull request now and, when everything is ready, create a different pull request to update the URLs to the operational ones (it should be an easy process). If you don't mind, I would like to implement your latest comments and include a message like 'not operational yet', just in case hehe. I will upload latest changes this Friday and then you can merge it, I will tell you here when everything is ready, is that ok? Thanks a lot!

bsipocz commented:
>  I will upload latest changes this Friday and then you can merge it, I will tell you here when everything is ready, is that ok? Thanks a lot!

Yes, I like that plan. I try to find the issue with the docbuild by then, as we have to have a green CI status for merging.

jespinosaar commented:
Hi @bsipocz , here you have the latest code with the changes, let's see if we can have the green CI status!

bsipocz commented:
@jespinosaar - I figured out the reason for the docs build failure, basically the config parser requires both config value and description. I added a some working ones (see my commit), please update them in the follow-up PR where you change the dummy URL to the real one.

I'll wait for CI now, and expect to see some more unrelated failures that have been fixed in the meantime on main, so then will go ahead and rebase this and fix any remaining issues.

bsipocz commented:
@pllim @saimn @astrofrog - this is a curious case of the testing suite namely if I run tests for the simbad submodule it all passes, but if it's run along with tests in all of ESA, two of them fail. I suspect this is some testing infrastructure related issue (we recently run into some cache related ones, etc.). 
Tests are still run with `python setup.py test -P esa,simbad`, but I'm fairly certain that's not part of the problem (and we don't see this same failure outside of this PR, could it be because of this JWST module uses Simbad itself?)

```
=============================================================================== test session starts ================================================================================
platform darwin -- Python 3.9.1, pytest-6.2.4, py-1.10.0, pluggy-0.13.1

Running tests with astroquery version 0.4.5.dev7217_testrun.
Running tests in astroquery/esa astroquery/simbad docs/esa docs/simbad.

Date: 2021-11-26T05:26:30

Platform: macOS-10.13.6-x86_64-i386-64bit

Executable: /Users/bsipocz/.pyenv/versions/3.9.1/bin/python

Full Python Version: 
3.9.1 (default, Jan 11 2021, 15:50:22) 
[Clang 10.0.0 (clang-1000.11.45.5)]

encodings: sys: utf-8, locale: UTF-8, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Package versions: 
Numpy: 1.19.5
Matplotlib: 3.3.3
Astropy: 4.3.1
APLpy: 2.0.3
pyregion: 2.1.1
regions: 0.4
pyVO: 1.1
mocpy: 0.8.5
astropy-healpix: 0.6
vamdclib: not available
astropy-helpers: 4.0.1

Using Astropy options: remote_data: none.

rootdir: /private/var/folders/dc/hsm7tqpx2d57n7vb3k1l81xw0000gq/T/astroquery-test-g55t4pj6/lib/python3.9/site-packages, configfile: setup.cfg
plugins: doctestplus-0.11.1, arraydiff-0.3, remotedata-0.3.2, hypothesis-6.3.1, filter-subpackage-0.1.1, openfiles-0.5.0, astropy-header-0.1.2, cov-3.0.0
collected 166 items                                                                                                                                                                

astroquery/esa/hubble/tests/test_esa_hubble.py ................                                                                                                              [  9%]
astroquery/esa/hubble/tests/test_esa_hubble_remote.py sssss                                                                                                                  [ 12%]
astroquery/esa/iso/tests/test_iso.py ....ss.....                                                                                                                             [ 19%]
astroquery/esa/jwst/tests/test_jwstdata.py .                                                                                                                                 [ 19%]
astroquery/esa/jwst/tests/test_jwsttap.py ...............                                                                                                                    [ 28%]
astroquery/esa/xmm_newton/tests/test_xmm_newton.py .................                                                                                                         [ 39%]
astroquery/esa/xmm_newton/tests/test_xmm_newton_remote.py ssssss                                                                                                             [ 42%]
astroquery/simbad/core.py ....                                                                                                                                               [ 45%]
astroquery/simbad/tests/test_simbad.py .............................................FF..........                                                                             [ 79%]
astroquery/simbad/tests/test_simbad_remote.py sssssssssssssssssssssssssssss                                                                                                  [ 96%]
docs/esa/hubble.rst s                                                                                                                                                        [ 97%]
docs/esa/iso.rst s                                                                                                                                                           [ 98%]
docs/esa/jwst.rst s                                                                                                                                                          [ 98%]
docs/esa/xmm_newton.rst s                                                                                                                                                    [ 99%]
docs/simbad/simbad.rst s                                                                                                                                                     [100%]

===================================================================================== FAILURES =====================================================================================
____________________________________________________________________________ test_query_object[m1-None] ____________________________________________________________________________

patch_post = <_pytest.monkeypatch.MonkeyPatch object at 0x1582f36d0>, object_name = 'm1', wildcard = None

    @pytest.mark.parametrize(('object_name', 'wildcard'),
                             [("m1", None),
                              ("m [0-9]", True),
                              ])
    def test_query_object(patch_post, object_name, wildcard):
        payload = simbad.core.Simbad.query_object(
            object_name, wildcard=wildcard, get_query_payload=True)
        expected_payload = {'script': '\nvotable {main_id,coordinates}\nvotable' +
                            ' open\nquery id {} {}  \nvotable close'.
                            format('wildcard' if wildcard else '', object_name)}
>       assert payload == expected_payload

astroquery/simbad/tests/test_simbad.py:324: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/Users/bsipocz/.pyenv/versions/3.9.1/lib/python3.9/site-packages/astropy/table/table.py:3361: in __eq__
    return self._rows_equal(other)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Table length=1>
MAIN_ID      RA         DEC     RA_PREC DEC_PREC COO_ERR_MAJA COO_ERR_MINA COO_ERR_ANGLE COO_QUAL COO... 31.94 +22 00 52.2       6        6           --           --             0        C              R 2011A&A...533A..10L
other = {'script': '\nvotable {main_id,coordinates}\nvotable open\nquery id  m1  \nvotable close'}

    def _rows_equal(self, other):
        """
        Row-wise comparison of table with any other object.
    
        This is actual implementation for __eq__.
    
        Returns a 1-D boolean numpy array showing result of row-wise comparison.
        This is the same as the ``==`` comparison for tables.
    
        Parameters
        ----------
        other : Table or DataFrame or ndarray
             An object to compare with table
    
        Examples
        --------
        Comparing one Table with other::
    
            >>> t1 = Table([[1,2],[4,5],[7,8]], names=('a','b','c'))
            >>> t2 = Table([[1,2],[4,5],[7,8]], names=('a','b','c'))
            >>> t1._rows_equal(t2)
            array([ True,  True])
    
        """
    
        if isinstance(other, Table):
            other = other.as_array()
    
        if self.has_masked_columns:
            if isinstance(other, np.ma.MaskedArray):
                result = self.as_array() == other
            else:
                # If mask is True, then by definition the row doesn't match
                # because the other array is not masked.
                false_mask = np.zeros(1, dtype=[(n, bool) for n in self.dtype.names])
>               result = (self.as_array().data == other) & (self.mask == false_mask)
E               FutureWarning: elementwise == comparison failed and returning scalar instead; this will raise an error or perform elementwise comparison in the future.

/Users/bsipocz/.pyenv/versions/3.9.1/lib/python3.9/site-packages/astropy/table/table.py:3401: FutureWarning
_________________________________________________________________________ test_query_object[m [0-9]-True] __________________________________________________________________________

patch_post = <_pytest.monkeypatch.MonkeyPatch object at 0x157d61a00>, object_name = 'm [0-9]', wildcard = True

    @pytest.mark.parametrize(('object_name', 'wildcard'),
                             [("m1", None),
                              ("m [0-9]", True),
                              ])
    def test_query_object(patch_post, object_name, wildcard):
        payload = simbad.core.Simbad.query_object(
            object_name, wildcard=wildcard, get_query_payload=True)
        expected_payload = {'script': '\nvotable {main_id,coordinates}\nvotable' +
                            ' open\nquery id {} {}  \nvotable close'.
                            format('wildcard' if wildcard else '', object_name)}
>       assert payload == expected_payload

astroquery/simbad/tests/test_simbad.py:324: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/Users/bsipocz/.pyenv/versions/3.9.1/lib/python3.9/site-packages/astropy/table/table.py:3361: in __eq__
    return self._rows_equal(other)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <Table length=1>
MAIN_ID      RA         DEC     RA_PREC DEC_PREC COO_ERR_MAJA COO_ERR_MINA COO_ERR_ANGLE COO_QUAL COO... 31.94 +22 00 52.2       6        6           --           --             0        C              R 2011A&A...533A..10L
other = {'script': '\nvotable {main_id,coordinates}\nvotable open\nquery id wildcard m [0-9]  \nvotable close'}

    def _rows_equal(self, other):
        """
        Row-wise comparison of table with any other object.
    
        This is actual implementation for __eq__.
    
        Returns a 1-D boolean numpy array showing result of row-wise comparison.
        This is the same as the ``==`` comparison for tables.
    
        Parameters
        ----------
        other : Table or DataFrame or ndarray
             An object to compare with table
    
        Examples
        --------
        Comparing one Table with other::
    
            >>> t1 = Table([[1,2],[4,5],[7,8]], names=('a','b','c'))
            >>> t2 = Table([[1,2],[4,5],[7,8]], names=('a','b','c'))
            >>> t1._rows_equal(t2)
            array([ True,  True])
    
        """
    
        if isinstance(other, Table):
            other = other.as_array()
    
        if self.has_masked_columns:
            if isinstance(other, np.ma.MaskedArray):
                result = self.as_array() == other
            else:
                # If mask is True, then by definition the row doesn't match
                # because the other array is not masked.
                false_mask = np.zeros(1, dtype=[(n, bool) for n in self.dtype.names])
>               result = (self.as_array().data == other) & (self.mask == false_mask)
E               FutureWarning: elementwise == comparison failed and returning scalar instead; this will raise an error or perform elementwise comparison in the future.

/Users/bsipocz/.pyenv/versions/3.9.1/lib/python3.9/site-packages/astropy/table/table.py:3401: FutureWarning
============================================================================= short test summary info ==============================================================================
FAILED astroquery/simbad/tests/test_simbad.py::test_query_object[m1-None] - FutureWarning: elementwise == comparison failed and returning scalar instead; this will raise an erro...
FAILED astroquery/simbad/tests/test_simbad.py::test_query_object[m [0-9]-True] - FutureWarning: elementwise == comparison failed and returning scalar instead; this will raise an...
==================================================================== 2 failed, 117 passed, 47 skipped in 3.30s =====================================================================
```

eerovaher commented:
Normally the `query_object()` method call at 
https://github.com/astropy/astroquery/blob/0833f634dbbfce9629b028192af9e1083e6b43fa/astroquery/simbad/tests/test_simbad.py#L319
leads first here:
https://github.com/astropy/astroquery/blob/0833f634dbbfce9629b028192af9e1083e6b43fa/astroquery/simbad/core.py#L543
and then here:
https://github.com/astropy/astroquery/blob/0833f634dbbfce9629b028192af9e1083e6b43fa/astroquery/simbad/core.py#L580
Simbad tests replace `SimbadClass._request()`:
https://github.com/astropy/astroquery/blob/0833f634dbbfce9629b028192af9e1083e6b43fa/astroquery/simbad/tests/test_simbad.py#L65
But the JWST tests replace the `Simbad.query_object()` method entirely:
https://github.com/astropy/astroquery/blob/e572ac20b9bc70f8f0f83851bc4176eda1037dfb/astroquery/esa/jwst/tests/test_jwsttap.py#L983
This means that the chain of method calls described above does not take place anymore and the monkeypatched result from the `query_object()` call is the one from the JWST tests.

The simplest fix would be to change the JWST tests so that they would create a new instance of `SimbadClass` so that `Simbad` from `astroquery.simbad` would not be affected by the monkeypatching in JWST tests:
```python
from astroquery.simbad import SimbadClass
Simbad = SimbadClass()
```

bsipocz commented:
Yes, indeed, thanks for the investigating. The issue only comes up for `get_query_payload` as that one returns sooner than the code reaches the newly patched `_request`. 

Anyway, I see two main points forward, that's beyond this PR:

- ideally, we should keep the module level tests separate, they should run in separate contexts. Not sure whether that's easily possible with pytest. @saimn would know I think.
- Make it better documented that classes are supposed to use as class instances. That will resolve some of the attribute related issues, too, etc. I would love to hear ideas of how to achieve that effect, when OO usage is not necessarily come naturally, and I think our docs is also a chaotic mix of inconsistencies on that front (and it may also not work some modules, which ideally should be also fixed).

jespinosaar commented:
Thanks a lot for your help on this @bsipocz and @eerovaher ! Well, we are progressing!

I think I can solve this error, as this is now an assertion in one of the messages. I have splitted the test_query_target test in two parts, the first one with the errors and the second one with the correct inputs. The first test is working with your fix, but the second one is trying to connect to internet... I have commented this second test, I think I should find a less complex way of testing this method.

@bsipocz. I can upload this fix now, what do you think?

bsipocz commented:
@jespinosaar - yes, please go ahead, I got side tracked into some cleanup tasks while we were waiting on CI :)

bsipocz commented on astroquery/esa/jwst/tests/test_jwsttap.py:
```diff
@@ -953,57 +953,61 @@ def test_query_target(self):
-            dummyTapHandler.check_call('launch_job', parameters)
+        assert 'This target name cannot be determined with this resolver: VIZIER' in err.value.args[0]
+
+    # def test_query_target(self):
```
Please do rather delete it than comment out, unless we're putting this back in once the remote tests are added with the functional archive

jespinosaar replied:
Ok, I am going to remove it.

jespinosaar commented:
Ok, now it seems everything is finally working!

bsipocz commented:
I've double-checked all the modules this uses, and run the remotes, too. `test_query_target_error` fails if I run the tests remotely, but it's OK as it's related to the dummy URL.  So let's go ahead and merge!

bsipocz commented:
Thank you @jespinosaar and everyone else involved!

jespinosaar commented:
Many thanks to everyone for your cooperation and support!! See you when we update the URL and release eJWST!

saimn commented:
> ideally, we should keep the module level tests separate, they should run in separate contexts. Not sure whether that's easily possible with pytest. @saimn would know I think.

@bsipocz - Unless you run tests for each modules in a separate process I don't think there is an easy solution.
